optimum/amd/brevitas/quantizer.py

```diff
@@ -0,0 +1,271 @@
+            model_name_or_path,
+            subfolder=subfolder,
+            revision=revision,
+            cache_dir=cache_dir,
+            use_auth_token=use_auth_token,
+            local_files_only=local_files_only,
+            force_download=force_download,
+            trust_remote_code=trust_remote_code,
+        )
+
+        return cls(model)
+
+    def quantize(
+        self, quantization_config: BrevitasQuantizationConfig, calibration_dataset: Optional[List[Dict]] = None
+    ) -> torch.nn.Module:
+        """
+        Quantizes the model using Brevitas according to the `quantization_config`.
+
+        Arguments:
+            quantization_config (`BrevitasQuantizationConfig`):
+                Quantization configuration to use to quantize the model.
+            calibration_dataset (`Optional[List[Dict]]`, defaults to `None`):
+                In case the quantization involves a calibration phase, this argument needs to be specified as a list of inputs to the model.
+                Example: `calibration_dataset = [{"input_ids": torch.tensor([[1, 2, 3, 4]])}, {"input_ids": torch.tensor([[6, 7, 3, 4]])}]` which is a dataset for a model taking `input_ids` as an argument, and which has two samples.
+        """
+        self.validate_quant_config(quantization_config)
+
+        requires_data = (
+            quantization_config.activations_equalization
+            or quantization_config.apply_gptq
+            or quantization_config.is_static
+        )
+        if calibration_dataset is None and requires_data:
+            raise ValueError(
+                f"No calibration_dataset was passed, but a calibration dataset is required with the quantization configuration activations_equalization={quantization_config.activations_equalization}, apply_gptq={quantization_config.apply_gptq}, is_static={quantization_config.is_static}."
+            )
+
+        if quantization_config.requires_fx_graph():
+            forward_signature = inspect.signature(self.model.forward).parameters
+            if all(
+                input_name in forward_signature for input_name in ["input_ids", "attention_mask", "past_key_values"]
+            ):
+                input_names = ["input_ids", "attention_mask", "past_key_values"]
+            else:
+                raise ValueError(
+                    f"Quantization with an FX graph is currently only supported for models taking `input_ids`, `attention_mask` and `past_key_values` as inputs. The model only has the following inputs: {forward_signature}"
+                )
+
+            with torch.no_grad():
+                model = symbolic_trace(self.model, input_names)
+        else:
+            model = self.model
+
+        dtype = next(iter(model.parameters())).dtype
+
+        # Insert standard MHA layers when performing fx based weight/act equalization to avoid dealing
+        # with all the variability in HF implementations
+        if quantization_config.replace_mha_with_quantizable:
```

Comment:
This replacement must be done before symbolic trace, otherwise we cannot recognise anymore a OPTAttention, since they are all nn.Module